Introducing DependencyViewer for querying Node's depedency relationships

SherlockNoMad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82906

This PR refactor the dependency map builder from the CapabilityBasedPartitioner into a standalone component. 

cc @ezyang @soumith @EikanWang @jgong5 @wenzhe-nrv

SherlockNoMad committed:
Introducing GraphViewer for querying Node's depedency relationships

[ghstack-poisoned]

davidberard98 commented on torch/fx/passes/infra/partitioner.py:
```diff
@@ -140,7 +99,7 @@ def assign(node: Node, id: Optional[int] = None):
                     pi = user_partitions_list[i]
                     pj = user_partitions_list[j]
-                    dependency = self.__partition_depends_on(pi, pj)
+                    dependency = self.graph_viewer.partition_depends_on(pi.nodes, pj.nodes)
```
for this specific case is it true that: for a two given inputs `pi.nodes` and `pj.nodes`, 1 and -1 cannot both be valid values of `dependency`?

SherlockNoMad replied:
yes.

SherlockNoMad commented on torch/fx/passes/utils/graph_viewer.py:
```diff
@@ -0,0 +1,58 @@
+        else:
+            return 0
+
+    def partition_depends_on(self, partition_a: Iterable[Node], partition_b: Iterable[Node]) -> int:
+        """
+        Returns
+            1 if b depends on a (,or equivalently a is an upstream depedency of b)
+            -1 if a depends on b (,or equivalently b is an upstream depedency of a)
+            0 if a and b doesn't have dependency between each other
+        """
```
1 and -1 cannot be valid at the same time, otherwise we have a loop in the graph. 
I can add documentation for this.

SherlockNoMad replied:
The other assumptions this function have is that 
- two partitions does share nodes
- partition itself is a connected graph

I guess I should assert this.

davidberard98 replied:
for arbitrary inputs a, b - you could have this?

```
         +------+
         |      |
       / +--+   |
      v     |   |
  +---+     |   |
  | A |     | B |
  +---+     |   |
    |       |   |
    +-----> |   |
            +---+
```

SherlockNoMad replied:
Addressed.

SherlockNoMad committed:
Update on "Introducing GraphViewer for querying Node's depedency relationships"


This PR refactor the dependency map builder from the CapabilityBasedPartitioner into a standalone component. 

[ghstack-poisoned]
